In [ ]:
import pandas as pd
import sklearn
import numpy as np
import matplotlib.pyplot as plt

In [2]:
# Grab and process the raw data.
imdb = pd.read_csv('imdb_labelled.txt', delimiter= '\t', header=None)
imdb.columns = ['review', 'positive']


In [3]:
keywords1 = ['best', 'adorable', 'loved', 'decent', 'terrific', 'liked', 'brilliant', 'rocked', 'great', 'fun', 'funny', '10']


keywords2 = ['best', 'adorable', 'loved', 'decent', 'terrific',
            'liked', 'brilliant', 'rocked', 'great', 'fun',
            'funny', '10', 'happy', 'worth', 'good', 'classic',
           'stars', 'treat', 'entertaining', 'smart', 'recommend', 
           'hot', 'wonderful', 'adventure', 'humor', 'humour', 'popular',
           'enjoyed', 'like', 'love', 'excellent', 'uplifting', 'amazed',
           'amazing', 'spectacular', 'mesmorizing','mesmerizing', 
            'excellent', 'balance', 'decent', 'rocked', 'sweet', 'pleased', 
           'awesome', 'powerful', 'solid', 'hilarious', 'identify', 'likes',
           'impressed', 'wonderful', 'appreciate', 'enjoy', 'enjoyed', 'recommended',
           'sincere', 'genuine', 'chemistry', 'natural', 'incredible', 'lovely', 
           'recommend', 'clever', 'hip', 'amazing']
keywords3 = ['best', 'adorable', 'loved', 'decent', 'terrific',
            'liked', 'brilliant', 'rocked', 'great', 'fun',
             '10', 'happy','classic',
            'recommend', 
           'hot', 'wonderful', 'adventure', 'humor', 'humour', 'popular',
           'enjoyed', 'like', 'love', 'excellent', 'uplifting', 'amazed',
           'amazing', 'spectacular', 'mesmorizing','mesmerizing', 
            'excellent', 'decent', 'rocked', 'pleased', 
           'awesome', 'powerful', 'solid', 'hilarious', 'identify', 'likes',
           'impressed', 'wonderful', 'appreciate', 'enjoy', 'enjoyed', 'recommended',
           'sincere', 'genuine', 'incredible', 'lovely', 
           'recommend', 'clever', 'amazing']
keywords4 = ['loved', 'great', 'wonderful', 'enjoyed', 'amazing', 
             'genuine', '10','classic']


In [37]:
keywords = keywords2
for key in keywords:
    imdb[str(key)] = imdb.review.str.contains(
        ' ' + str(key) + ' ',
        case=False
    )

In [38]:
imdb['positive'] = (imdb['positive'] == 1)

In [39]:
data = imdb[keywords]
target = imdb['positive']

In [40]:
# Our data is binary / boolean, so we're importing the Bernoulli classifier.
from sklearn.naive_bayes import BernoulliNB
bnb = BernoulliNB()
y_pred = bnb.fit(data, target).predict(data)

print("Number of mislabeled points out of a total of {} points : {}".format(
     data.shape[0],
     (target != y_pred).sum()
))

Number of mislabeled points out of a total of 748 points : 245


In [41]:
# Test your model with different holdout groups.

from sklearn.model_selection import train_test_split
# Use train_test_split to create the necessary training and test groups
X_train, X_test, y_train, y_test = train_test_split(data, target, test_size=0.2, random_state=20)
print('With 20% Holdout: ' + str(bnb.fit(X_train, y_train).score(X_test, y_test)))
print('Testing on Sample: ' + str(bnb.fit(data, target).score(data, target)))

With 20% Holdout: 0.58
Testing on Sample: 0.6724598930481284


In [42]:
from sklearn.model_selection import cross_val_score
cross_val_score(bnb, data, target, cv=10)

array([0.59210526, 0.67105263, 0.70666667, 0.6       , 0.69333333,
       0.6       , 0.72972973, 0.59459459, 0.62162162, 0.60810811])

In [43]:
from sklearn import datasets, linear_model
from sklearn.model_selection import train_test_split
X1 = data
y1 = target
X1_train, X1_test, y1_train, y1_test = train_test_split(X1, y1, test_size=0.2)
print(X1_train.shape, y1_train.shape)
print(X1_test.shape, y1_test.shape)

(598, 65) (598,)
(150, 65) (150,)


In [44]:
from sklearn.naive_bayes import BernoulliNB
#Cross validation model
from sklearn.model_selection import KFold # import KFold
kf = KFold(n_splits = 5, shuffle=True, random_state = 40)
bnb = BernoulliNB() 
for train_index, test_index in kf.split(X1):
    X1_train, X1_test = X1.iloc[train_index], X1.iloc[test_index]
    y1_train, y1_test = y1.iloc[train_index], y1.iloc[test_index]
    bnb = bnb.fit(X1_train, y1_train)
    print("Score:", bnb.score(X1_test, y1_test))

Score: 0.6733333333333333
Score: 0.6133333333333333
Score: 0.6333333333333333
Score: 0.6711409395973155
Score: 0.5973154362416108


Keywords1: Number of mislabeled points out of a total of 748 points : 324

Keywords2: Number of mislabeled points out of a total of 748 points : 245

Keywords3: Number of mislabeled points out of a total of 748 points : 265

Keywords3: Number of mislabeled points out of a total of 748 points : 335
